### ISSUE TITLE
help with Kullback–Leibler divergence loss

### ISSUE THREAD

**Issue by schmolze** [fchollet/keras#6183]
I'm training a network on survival curves. The curves cover 20 years, and I'm generating data at 6 month intervals, so my targets are 40 element vectors with each value representing the probability of survival at that time period.

I'd like to minimize the Kullback–Leibler divergence between the final output layer (40 units) and the target survival curve. My data is gene expression data for 87 genes, and I have 1088 cases, so my input shape is (1088, 87) and my target shape is (1088, 40). Here is my model:

`model = Sequential()`

`model.add(Dense(64, activation='relu', input_dim=input_dim))`
`model.add(Dropout(0.5))`
`model.add(Dense(64, activation='relu'))`
`model.add(Dropout(0.5))`
`model.add(Dense(40, activation='sigmoid'))`

`sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)`

`model.compile(loss='kld', optimizer=sgd, metrics=['accuracy'])`

`model.fit(X, Y, validation_split=0.33, epochs=200, [batch_size=64)`]

And here's what the accuracy and loss look like:

![acc](https://cloud.githubusercontent.com/assets/3537118/24770019/5cf0fc4a-1abd-11e7-8cf3-4a568469c172.png)

Val accuracy approaching 100%, but train accuracy less?

![loss](https://cloud.githubusercontent.com/assets/3537118/24770049/70ef521e-1abd-11e7-9b1b-95cc13fdc02e.png)

Negative loss?

If I try predicting from new input data with

`model.predict(X_new)`

I get all 1's, or values like 0.99999988.

Any suggestions would be appreciated.

**Comment by Ultramann** [comment]
@schmolze Fast guess, since you're activating with sigmoid for your predictions each one of the nodes will be outputting a value between 0 and 1. This kind of sounds like what you want, however you run into a problem when you use KL divergence as your loss function since it expects to compare two proper distributions. If you want to propagate loss per time period then I think the right way to do it is with binary crossentropy, if you in fact want a proper distribution to use with KL divergence then the you should activate with softmax on the last layer.

Final thought, how many total observations are you training vs testing with after data augmentation, you might simply be seeing the result of a poor train test split in those graphs if your data are small. Just some food for thought.

**Comment by bstriner** [comment]
What are the target values? Are they probabilities or alive/dead states? If alive/dead then @Ultramann is right and you want binary crossentropy.

The output predictions are not independent. I assume if someone is dead at one state they are likely to be dead in the next state. You may want to make predictions with some sort of recurrency.

**Comment by schmolze** [comment]
Thanks for the input guys. I've given the problem more thought, and I think I have a working solution now. First, a bit more on the survival curves, which are based on a 1999 paper by Street entitled "A Neural Network Model for Prognostic Prediction". If a patient is known to be alive at a given time point, that index gets a 1, if they are known to be dead a 0. If they are censored (e.g. lost to followup), the probability of survival is estimated with the Kaplan-Meier estimator. So, example target vectors might look like:

`[1,1,1,0,0,0,.....,0]`
`[1,1,0.98,0.95,0.94,....0.25]`

Since as pointed out, these aren't actual probability distributions, I don't think K-L is the right loss function. Instead, I ended up just treating this as a standard regression problem, and used MSE loss. After some tweaks to the architecture, it seems to be working pretty well:

![figure_1](https://cloud.githubusercontent.com/assets/3537118/24885772/3c0eadba-1e05-11e7-872d-b7b18d163330.png)
